File: crates/countersign-verify/src/jcs.rs
//! RFC 8785 (JSON Canonicalization Scheme), restricted to integers.
//!
//! Canonicalization is the step where two implementations disagree silently and
//! a valid approval fails to verify. So this is written out rather than pulled
//! in, and every rule that differs from "serialize the JSON" carries the reason
//! it exists.
//!
//! # The integer restriction
//!
//! RFC 8785 serializes numbers with ECMAScript `Number::toString` — a
//! shortest-round-trip double formatter. It is well-defined, and it is a
//! reliable source of cross-language divergence, which in this layer means a
//! human turned the dial and the verifier said no.
//!
//! Every number Countersign carries is a count, a millisecond, or a version. So
//! [`canonicalize`] **rejects** any number that is not an integer in the
//! JavaScript-safe range rather than serializing it approximately. Within that
//! subset the output is byte-identical to RFC 8785, so this restricts inputs
//! without diverging from the format.

use std::cmp::Ordering;
use std::fmt::Write as _;

use serde_json::Value;

/// The largest integer representable exactly in an IEEE-754 double, and so the
/// largest one every JSON implementation agrees about.
const SAFE_INT_MAX: i128 = 9_007_199_254_740_991; // 2^53 - 1

/// Why a value could not be canonicalized.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum JcsError {
    /// A number was not an integer, or fell outside ±(2^53 - 1).
    UnsupportedNumber(String),
    /// An object carried the same key twice.
    DuplicateKey(String),
    /// The input was not well-formed JSON.
    Malformed(String),
}

impl std::fmt::Display for JcsError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            JcsError::UnsupportedNumber(n) => write!(
                f,
                "number {n} is not an integer within ±(2^53-1); Countersign v1 canonicalizes integers only"
            ),
            JcsError::DuplicateKey(k) => write!(f, "duplicate object key {k:?}"),
            JcsError::Malformed(e) => write!(f, "malformed JSON: {e}"),
        }
    }
}

impl std::error::Error for JcsError {}

/// Canonicalize a parsed value.
///
/// Duplicate keys cannot be detected here — `serde_json::Value` has already
/// collapsed them. Use [`canonicalize_str`] when the input arrived as text from
/// somewhere you do not control.
pub fn canonicalize(value: &Value) -> Result<String, JcsError> {
    let mut out = String::new();
    write_value(value, &mut out)?;
    Ok(out)
}

/// Canonicalize JSON text, rejecting duplicate object keys.
///
/// RFC 8785 requires duplicate rejection, and it is not pedantry here: two
/// parsers that disagree about which of `{"ttl_ms":1,"ttl_ms":999999}` wins
/// produce different digests for the same bytes, and the disagreement is
/// attacker-chosen.
pub fn canonicalize_str(json: &str) -> Result<String, JcsError> {
    // serde_json keeps the last duplicate silently, so the check has to happen
    // during parsing rather than after it.
    let mut de = serde_json::Deserializer::from_str(json);
    let checked: DupChecked =
        serde::Deserialize::deserialize(&mut de).map_err(|e| JcsError::Malformed(e.to_string()))?;
    de.end().map_err(|e| JcsError::Malformed(e.to_string()))?;
    if let Some(k) = checked.duplicate {
        return Err(JcsError::DuplicateKey(k));
    }
    canonicalize(&checked.value)
}

/// A `Value` plus the first duplicate key seen while parsing it.
struct DupChecked {
    value: Value,
    duplicate: Option<String>,
}

impl<'de> serde::Deserialize<'de> for DupChecked {
    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
        d.deserialize_any(DupVisitor)
    }
}

struct DupVisitor;

impl<'de> serde::de::Visitor<'de> for DupVisitor {
    type Value = DupChecked;

    fn expecting(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("any JSON value")
    }

    fn visit_unit<E>(self) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::Null,
            duplicate: None,
        })
    }
    fn visit_none<E>(self) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::Null,
            duplicate: None,
        })
    }
    fn visit_bool<E>(self, v: bool) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::Bool(v),
            duplicate: None,
        })
    }
    fn visit_i64<E>(self, v: i64) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::from(v),
            duplicate: None,
        })
    }
    fn visit_u64<E>(self, v: u64) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::from(v),
            duplicate: None,
        })
    }
    fn visit_f64<E>(self, v: f64) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::from(v),
            duplicate: None,
        })
    }
    fn visit_str<E>(self, v: &str) -> Result<DupChecked, E> {
        Ok(DupChecked {
            value: Value::String(v.to_owned()),
            duplicate: None,
        })
    }

    fn visit_seq<A: serde::de::SeqAccess<'de>>(self, mut a: A) -> Result<DupChecked, A::Error> {
        let mut items = Vec::new();
        let mut duplicate = None;
        while let Some(el) = a.next_element::<DupChecked>()? {
            duplicate = duplicate.or(el.duplicate);
            items.push(el.value);
        }
        Ok(DupChecked {
            value: Value::Array(items),
            duplicate,
        })
    }

    fn visit_map<A: serde::de::MapAccess<'de>>(self, mut a: A) -> Result<DupChecked, A::Error> {
        let mut map = serde_json::Map::new();
        let mut duplicate = None;
        while let Some(k) = a.next_key::<String>()? {
            let v = a.next_value::<DupChecked>()?;
            duplicate = duplicate.or(v.duplicate);
            if map.insert(k.clone(), v.value).is_some() && duplicate.is_none() {
                duplicate = Some(k);
            }
        }
        Ok(DupChecked {
            value: Value::Object(map),
            duplicate,
        })
    }
}

fn write_value(v: &Value, out: &mut String) -> Result<(), JcsError> {
    match v {
        Value::Null => out.push_str("null"),
        Value::Bool(true) => out.push_str("true"),
        Value::Bool(false) => out.push_str("false"),
        Value::Number(n) => write_number(n, out)?,
        Value::String(s) => write_string(s, out),
        Value::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_value(item, out)?;
            }
            out.push(']');
        }
        Value::Object(map) => {
            // Sorted by UTF-16 code unit, which is neither byte order nor code
            // point order — see `utf16_cmp`.
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort_by(|a, b| utf16_cmp(a, b));
            out.push('{');
            for (i, k) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_string(k, out);
                out.push(':');
                write_value(&map[*k], out)?;
            }
            out.push('}');
        }
    }
    Ok(())
}

fn write_number(n: &serde_json::Number, out: &mut String) -> Result<(), JcsError> {
    let as_int: Option<i128> = n
        .as_i64()
        .map(i128::from)
        .or_else(|| n.as_u64().map(i128::from));

    match as_int {
        Some(i) if i.abs() <= SAFE_INT_MAX => {
            let _ = write!(out, "{i}");
            Ok(())
        }
        _ => Err(JcsError::UnsupportedNumber(n.to_string())),
    }
}

/// RFC 8785 string escaping: only `"`, `\` and C0 controls are escaped, the
/// five short forms are used where they exist, other controls take `\u00xx`
/// with **lowercase** hex, `/` is left alone, and non-ASCII is emitted as UTF-8
/// rather than `\u`-escaped.
fn write_string(s: &str, out: &mut String) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{08}' => out.push_str("\\b"),
            '\u{09}' => out.push_str("\\t"),
            '\u{0a}' => out.push_str("\\n"),
            '\u{0c}' => out.push_str("\\f"),
            '\u{0d}' => out.push_str("\\r"),
            c if (c as u32) < 0x20 => {
                let _ = write!(out, "\\u{:04x}", c as u32);
            }
            c => out.push(c),
        }
    }
    out.push('"');
}

/// Compare two strings by UTF-16 code unit, as RFC 8785 requires.
///
/// Rust's `str: Ord` compares UTF-8 bytes, which orders by code point. The two
/// disagree above the BMP: a code point at or above `U+10000` encodes in UTF-16
/// as a surrogate pair beginning in `0xD800..=0xDBFF`, so it sorts *before*
/// `U+E000..=U+FFFF` — the reverse of code-point order. Object keys are usually
/// ASCII and this rarely bites, which is exactly why it needs a test rather
/// than a comment.
fn utf16_cmp(a: &str, b: &str) -> Ordering {
    a.encode_utf16().cmp(b.encode_utf16())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn canon(v: Value) -> String {
        canonicalize(&v).expect("canonicalizable")
    }

    #[test]
    fn object_keys_sort_and_whitespace_vanishes() {
        assert_eq!(canon(json!({"b": 1, "a": 2})), r#"{"a":2,"b":1}"#);
    }

    #[test]
    fn keys_sort_by_utf16_code_unit_not_code_point() {
        // U+FFFD sorts *after* U+10140 in UTF-16 (the latter starts with a
        // 0xD800-range surrogate) and *before* it by code point. Byte-wise
        // sorting — Rust's default — would produce the wrong order here.
        let high = "\u{10140}"; // surrogate pair D800 DD40
        let bmp = "\u{FFFD}";
        let out = canon(json!({ bmp: 1, high: 2 }));
        let expected = format!("{{\"{high}\":2,\"{bmp}\":1}}");
        assert_eq!(out, expected, "keys must sort by UTF-16 code unit");
        assert!(bmp < high, "and that is the opposite of Rust's byte order");
    }

    #[test]
    fn strings_escape_exactly_what_rfc8785_escapes() {
        assert_eq!(canon(json!("a/b")), r#""a/b""#, "solidus is not escaped");
        assert_eq!(canon(json!("q\"s")), r#""q\"s""#);
        assert_eq!(canon(json!("back\\slash")), r#""back\\slash""#);
        assert_eq!(canon(json!("tab\there")), r#""tab\there""#);
        assert_eq!(canon(json!("nl\n")), r#""nl\n""#);
        // A control with no short form takes lowercase-hex \u00xx.
        assert_eq!(canon(json!("\u{01}")), "\"\\u0001\"");
        // Non-ASCII travels as UTF-8, not as an escape.
        assert_eq!(canon(json!("café")), "\"café\"");
    }

    #[test]
    fn non_integer_numbers_are_refused_rather_than_approximated() {
        assert!(matches!(
            canonicalize(&json!(1.5)),
            Err(JcsError::UnsupportedNumber(_))
        ));
        // 2^53 is the first integer two doubles can disagree about.
        assert!(matches!(
            canonicalize(&json!(9_007_199_254_740_992i64)),
            Err(JcsError::UnsupportedNumber(_))
        ));
        assert_eq!(canon(json!(9_007_199_254_740_991i64)), "9007199254740991");
        assert_eq!(canon(json!(-9_007_199_254_740_991i64)), "-9007199254740991");
        assert_eq!(canon(json!(0)), "0");
    }

    #[test]
    fn duplicate_keys_are_rejected_from_text() {
        // Whichever key a parser keeps, it is a different digest for the same
        // bytes — and the choice is the sender's.
        assert!(matches!(
            canonicalize_str(r#"{"ttl_ms":1,"ttl_ms":999999}"#),
            Err(JcsError::DuplicateKey(k)) if k == "ttl_ms"
        ));
        // Nested, and inside an array, too.
        assert!(matches!(
            canonicalize_str(r#"{"a":{"x":1,"x":2}}"#),
            Err(JcsError::DuplicateKey(_))
        ));
        assert!(matches!(
            canonicalize_str(r#"[{"x":1,"x":2}]"#),
            Err(JcsError::DuplicateKey(_))
        ));
        assert!(canonicalize_str(r#"{"a":1,"b":2}"#).is_ok());
    }

    #[test]
    fn nesting_and_arrays_round_trip() {
        assert_eq!(
            canon(json!({"z": [1, {"b": true, "a": null}], "y": "x"})),
            r#"{"y":"x","z":[1,{"a":null,"b":true}]}"#
        );
        assert_eq!(canon(json!([])), "[]");
        assert_eq!(canon(json!({})), "{}");
    }

    #[test]
    fn canonical_form_is_a_fixed_point() {
        let once = canonicalize_str(r#" { "b" : 1 , "a" : [ 2 , 3 ] } "#).unwrap();
        assert_eq!(once, r#"{"a":[2,3],"b":1}"#);
        assert_eq!(canonicalize_str(&once).unwrap(), once);
    }
}

File: sdk/typescript/src/jcs.ts
/**
 * RFC 8785 (JSON Canonicalization Scheme), restricted to integers.
 *
 * A direct port of `countersign-verify`'s `jcs.rs`. Canonicalization is where
 * two implementations disagree silently and a valid approval fails to verify,
 * so this is written out rather than depended on, and checked against the same
 * committed vectors the Rust one is.
 *
 * Two things are easier here than in Rust, and one is harder.
 *
 * **Easier:** JavaScript strings are already UTF-16, so the default `sort()`
 * comparison is UTF-16 code-unit order — exactly what RFC 8785 requires. In
 * Rust that had to be written out, because `str: Ord` compares UTF-8 bytes and
 * the two disagree above the BMP.
 *
 * **Also easier:** `Number.isInteger` and `Number.MAX_SAFE_INTEGER` are the
 * spec's own vocabulary for the restriction this profile applies.
 *
 * **Harder:** `JSON.parse` collapses duplicate keys silently, so detecting them
 * needs a separate scan of the source text. See {@link findDuplicateKey}.
 */

/** The largest integer every JSON implementation agrees about: 2^53 - 1. */
const SAFE_INT_MAX = Number.MAX_SAFE_INTEGER;

export type JcsErrorKind = "unsupported_number" | "duplicate_key" | "malformed";

export class JcsError extends Error {
  readonly kind: JcsErrorKind;

  constructor(message: string, kind: JcsErrorKind) {
    super(message);
    this.name = "JcsError";
    this.kind = kind;
  }
}

/**
 * Canonicalize an already-parsed value.
 *
 * Duplicate keys cannot be detected here — the parse has already collapsed
 * them. Use {@link canonicalizeText} for input from somewhere you do not
 * control.
 */
export function canonicalize(value: unknown): string {
  const out: string[] = [];
  writeValue(value, out);
  return out.join("");
}

/** Canonicalize JSON text, rejecting duplicate object keys. */
export function canonicalizeText(json: string): string {
  const duplicate = findDuplicateKey(json);
  if (duplicate !== null) {
    throw new JcsError(`duplicate object key ${JSON.stringify(duplicate)}`, "duplicate_key");
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch (e) {
    throw new JcsError(`malformed JSON: ${(e as Error).message}`, "malformed");
  }
  return canonicalize(parsed);
}

function writeValue(value: unknown, out: string[]): void {
  if (value === null) {
    out.push("null");
    return;
  }
  switch (typeof value) {
    case "boolean":
      out.push(value ? "true" : "false");
      return;
    case "number":
      writeNumber(value, out);
      return;
    case "string":
      out.push(writeString(value));
      return;
    case "object":
      break;
    default:
      throw new JcsError(`cannot canonicalize a ${typeof value}`, "malformed");
  }

  if (Array.isArray(value)) {
    out.push("[");
    value.forEach((item, index) => {
      if (index > 0) out.push(",");
      writeValue(item, out);
    });
    out.push("]");
    return;
  }

  // Sorted by UTF-16 code unit, which is what the default comparison already
  // does — JavaScript strings are UTF-16. This is the rule that has to be
  // written out by hand in languages whose strings are not.
  const keys = Object.keys(value as Record<string, unknown>).sort();
  out.push("{");
  keys.forEach((key, index) => {
    if (index > 0) out.push(",");
    out.push(writeString(key));
    out.push(":");
    writeValue((value as Record<string, unknown>)[key], out);
  });
  out.push("}");
}

function writeNumber(value: number, out: string[]): void {
  if (!Number.isInteger(value) || Math.abs(value) > SAFE_INT_MAX) {
    throw new JcsError(
      `number ${value} is not an integer within ±(2^53-1); Countersign v1 canonicalizes ` +
        `integers only`,
      "unsupported_number",
    );
  }
  // `-0` would serialize as "0" via String(), which is what we want, but be
  // explicit rather than relying on it.
  out.push(String(value === 0 ? 0 : value));
}

/**
 * RFC 8785 escaping: only `"`, `\` and C0 controls, the five short forms where
 * they exist, `\u00xx` with lowercase hex otherwise. `/` is not escaped, and
 * non-ASCII travels as UTF-8.
 */
function writeString(value: string): string {
  let out = '"';
  for (const char of value) {
    switch (char) {
      case '"':
        out += '\\"';
        break;
      case "\\":
        out += "\\\\";
        break;
      case "\b":
        out += "\\b";
        break;
      case "\t":
        out += "\\t";
        break;
      case "\n":
        out += "\\n";
        break;
      case "\f":
        out += "\\f";
        break;
      case "\r":
        out += "\\r";
        break;
      default: {
        const code = char.codePointAt(0)!;
        if (code < 0x20) {
          out += "\\u" + code.toString(16).padStart(4, "0");
        } else {
          out += char;
        }
      }
    }
  }
  return out + '"';
}

/**
 * Find the first duplicate object key in JSON text, or `null`.
 *
 * `JSON.parse` keeps the last of a duplicate pair and says nothing, so two
 * parsers can disagree about which of `{"ttl_ms":1,"ttl_ms":999999}` wins and
 * produce different digests for the same bytes — with the disagreement chosen
 * by whoever sent them. RFC 8785 requires rejection, so this scans the source.
 *
 * It is a scanner, not a parser: it tracks nesting and string state well enough
 * to know which tokens are keys, and leaves every other judgement to
 * `JSON.parse`.
 */
export function findDuplicateKey(json: string): string | null {
  const scopes: Array<Set<string> | null> = []; // null marks an array scope
  let i = 0;
  let expectKey = false;

  while (i < json.length) {
    const char = json[i];

    if (char === '"') {
      const [text, next] = readString(json, i);
      i = next;
      if (expectKey && scopes.length > 0) {
        const scope = scopes[scopes.length - 1];
        if (scope !== null) {
          if (scope.has(text)) return text;
          scope.add(text);
        }
        expectKey = false;
      }
      continue;
    }

    switch (char) {
      case "{":
        scopes.push(new Set());
        expectKey = true;
        break;
      case "[":
        scopes.push(null);
        expectKey = false;
        break;
      case "}":
      case "]":
        scopes.pop();
        expectKey = false;
        break;
      case ",":
        // A comma inside an object introduces the next key.
        expectKey = scopes.length > 0 && scopes[scopes.length - 1] !== null;
        break;
      default:
        break;
    }
    i += 1;
  }
  return null;
}

/** Read a JSON string starting at `start`, returning its value and end index. */
function readString(json: string, start: number): [string, number] {
  let i = start + 1;
  let raw = '"';
  while (i < json.length) {
    const char = json[i];
    raw += char;
    if (char === "\\") {
      // Skip the escaped character so an escaped quote does not end the string.
      raw += json[i + 1] ?? "";
      i += 2;
      continue;
    }
    i += 1;
    if (char === '"') break;
  }
  let text: string;
  try {
    text = JSON.parse(raw) as string;
  } catch {
    // Malformed; let JSON.parse report it properly later.
    text = raw;
  }
  return [text, i];
}
